File: database/database-annotation-processor/src/main/java/ru/tinkoff/kora/database/annotation/processor/r2dbc/R2dbcStatementSetterGenerator.java
package ru.tinkoff.kora.database.annotation.processor.r2dbc;

import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeName;
import ru.tinkoff.kora.annotation.processor.common.CommonUtils;
import ru.tinkoff.kora.annotation.processor.common.FieldFactory;
import ru.tinkoff.kora.database.annotation.processor.QueryWithParameters;
import ru.tinkoff.kora.database.annotation.processor.model.QueryParameter;

import javax.annotation.Nullable;
import javax.lang.model.element.ExecutableElement;
import java.util.List;
import java.util.Objects;

public class R2dbcStatementSetterGenerator {

    public static void generate(MethodSpec.Builder b, ExecutableElement method, QueryWithParameters sqlWithParameters, List<QueryParameter> parameters, @Nullable QueryParameter batchParam, FieldFactory parameterMappers) {
        if (batchParam != null) {
            b.addCode("""
                for (int i = 0; i < $L.size(); i++) {
                  var _batch_$L = $L.get(i);$>
                """, batchParam.name(), batchParam.name(), batchParam.name());
        }

        for (int i = 0, sqlIndex = 1; i < parameters.size(); i++, sqlIndex++) {
            var parameter = parameters.get(i);
            if (parameter instanceof QueryParameter.ConnectionParameter) {
                continue;
            }

            var parameterName = parameter.name();
            if (parameter instanceof QueryParameter.BatchParameter batchParameter) {
                parameter = batchParameter.parameter();
                parameterName = "_batch_" + parameter.name();
            }

            if (parameter instanceof QueryParameter.SimpleParameter simpleParameter) {
                var sqlParameter = Objects.requireNonNull(sqlWithParameters.find(i));
                var nativeType = R2dbcNativeTypes.findAndBox(TypeName.get(simpleParameter.type()));
                var mapping = CommonUtils.parseMapping(simpleParameter.variable()).getMapping(R2dbcTypes.PARAMETER_COLUMN_MAPPER);
                if (nativeType != null && mapping == null) {
                    for (var index : sqlParameter.sqlIndexes()) {
                        if (CommonUtils.isNullable(simpleParameter.variable())) {
                            b.addCode("""
                                if($L == null) {
                                  _stmt.bindNull($L, $L.class);
                                } else {
                                  _stmt.bind($L, $L);
                                }
                                """, parameterName, index, nativeType, index, parameterName);
                        } else {
                            b.addCode("_stmt.bind($L, $L);\n", index, parameterName);
                        }
                    }
                } else if (mapping != null && mapping.mapperClass() != null) {
                    var mapper = parameterMappers.get(mapping.mapperClass(), mapping.mapperTags());
                    for (var index : sqlParameter.sqlIndexes()) {
                        b.addCode("$L.apply(_stmt, $L, $L);\n", mapper, index, parameterName);
                    }
                } else {
                    var mapper = parameterMappers.get(R2dbcTypes.PARAMETER_COLUMN_MAPPER, simpleParameter.type(), simpleParameter.variable());
                    for (var index : sqlParameter.sqlIndexes()) {
                        b.addCode("$L.apply(_stmt, $L, $L);\n", mapper, index, parameterName);
                    }
                }
            }

            if (parameter instanceof QueryParameter.EntityParameter entityParameter) {
                for (var field : entityParameter.entity().columns()) {
                    var fieldAccessor = field.accessor();

                    var sqlParameter = sqlWithParameters.find(field.queryParameterName(entityParameter.name()));
                    if (sqlParameter == null || sqlParameter.sqlIndexes().isEmpty()) {
                        continue;
                    }
                    var nativeType = R2dbcNativeTypes.findAndBox(TypeName.get(field.type()));
                    var mapping = CommonUtils.parseMapping(field.element()).getMapping(R2dbcTypes.PARAMETER_COLUMN_MAPPER);
                    if (nativeType != null && mapping == null) {
                        for (var index : sqlParameter.sqlIndexes()) {
                            if (CommonUtils.isNullable(field.element())) {
                                b.addCode("""
                                    if($L == null) {
                                      _stmt.bindNull($L, $L.class);
                                    } else {
                                      _stmt.bind($L, $L);
                                    }
                                    """, fieldAccessor, index, nativeType, index, fieldAccessor);
                            } else {
                                b.addCode("_stmt.bind($L, $L);\n", index, fieldAccessor);
                            }
                        }
                    } else if (mapping != null && mapping.mapperClass() != null) {
                        var mapper = parameterMappers.get(mapping.mapperClass(), mapping.mapperTags());
                        for (var index : sqlParameter.sqlIndexes()) {
                            b.addCode("$L.apply(_stmt, $L, $L);\n", mapper, index, fieldAccessor);
                        }
                    } else {
                        var mapper = parameterMappers.get(R2dbcTypes.PARAMETER_COLUMN_MAPPER, field.type(), field.element());
                        for (var index : sqlParameter.sqlIndexes()) {
                            b.addCode("$L.apply(_stmt, $L, $N.$N());\n", mapper, index, parameterName, fieldAccessor);
                        }
                    }
                }
            }
        }

        if (batchParam != null) {
            b.addCode("""
                if(i != $L.size() - 1) {
                  _stmt.add();
                }""", batchParam.name());
            b.addCode("\n$<}\n");
        }
    }
}

File: database/database-symbol-processor/src/main/kotlin/ru/tinkoff/kora/database/symbol/processor/r2dbc/R2dbcStatementSetterGenerator.kt
package ru.tinkoff.kora.database.symbol.processor.r2dbc

import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.ksp.toTypeName
import ru.tinkoff.kora.database.symbol.processor.QueryWithParameters
import ru.tinkoff.kora.database.symbol.processor.model.QueryParameter
import ru.tinkoff.kora.ksp.common.FieldFactory
import ru.tinkoff.kora.ksp.common.KotlinPoetUtils.controlFlow
import ru.tinkoff.kora.ksp.common.parseMappingData

object R2dbcStatementSetterGenerator {
    fun generate(
        b: FunSpec.Builder,
        query: QueryWithParameters,
        parameters: List<QueryParameter>,
        batchParam: QueryParameter?,
        parameterMappers: FieldFactory
    ) {
        if (batchParam != null) {
            b.beginControlFlow("for (_batch_%L in %N)", batchParam.name, batchParam.name)
        }
        var sqlIndex = 0
        parameters.forEach { p ->
            var parameter = p
            if (parameter is QueryParameter.ConnectionParameter) {
                return@forEach
            }
            var parameterName = parameter.name
            if (parameter is QueryParameter.BatchParameter) {
                parameter = parameter.parameter
                parameterName = "_batch_${parameterName}"
            }
            if (parameter is QueryParameter.SimpleParameter) {
                val nativeType = R2dbcNativeTypes.findNativeType(parameter.type.toTypeName())
                val mapping = parameter.variable.parseMappingData().getMapping(R2dbcTypes.parameterColumnMapper)
                if (nativeType != null && mapping == null) {
                    if (parameter.type.isMarkedNullable) {
                        b.controlFlow("if (%L != null)", parameterName) {
                            addCode(nativeType.bind("_stmt", parameterName, sqlIndex)).addCode("\n")
                            nextControlFlow("else")
                            addCode(nativeType.bindNull("_stmt", sqlIndex)).addCode("\n")
                        }
                    } else {
                        b.addCode(nativeType.bind("_stmt", parameterName, sqlIndex)).addCode("\n")
                    }
                } else if (mapping?.mapper != null) {
                    val mapper = parameterMappers.get(mapping.mapper!!, mapping.tags)
                    b.addCode("%N.apply(_stmt, %L, %N)\n", mapper, sqlIndex, parameterName)
                } else {
                    val mapper = parameterMappers.get(R2dbcTypes.parameterColumnMapper, parameter.type, parameter.variable)
                    b.addCode("%N.apply(_stmt, %L, %N)\n", mapper, sqlIndex, parameterName)
                }
                sqlIndex++
                return@forEach
            }
            if (parameter is QueryParameter.EntityParameter) {
                for (field in parameter.entity.fields) {
                    val sqlParam = query.find(parameter.name + "." + field.property.simpleName.getShortName())
                    if (sqlParam?.sqlIndexes.isNullOrEmpty()) {
                        continue
                    }
                    val nativeType = R2dbcNativeTypes.findNativeType(field.type.toTypeName())
                    val fieldValue = if (parameter.type.isMarkedNullable) {
                        parameterName + "?." + field.property.simpleName.getShortName()
                    } else {
                        parameterName + "." + field.property.simpleName.getShortName()
                    }
                    val mapping = field.mapping.getMapping(R2dbcTypes.parameterColumnMapper)
                    if (nativeType != null && mapping == null) {
                        if (parameter.type.isMarkedNullable || field.type.isMarkedNullable) {
                            b.beginControlFlow("if (%L != null)", fieldValue)
                        }
                        b.addCode(nativeType.bind("_stmt", fieldValue, sqlIndex)).addCode("\n")
                        if (parameter.type.isMarkedNullable || field.type.isMarkedNullable) {
                            b.nextControlFlow("else")
                            b.addCode(nativeType.bindNull("_stmt", sqlIndex)).addCode("\n")
                            b.endControlFlow()
                        }
                    } else if (mapping?.mapper != null) {
                        val mapper = parameterMappers.get(mapping.mapper!!, mapping.tags)
                        b.addStatement("%N.apply(_stmt, %L, %L)", mapper, sqlIndex, fieldValue)
                    } else {
                        val mapper = parameterMappers.get(R2dbcTypes.parameterColumnMapper, field.type, field.property)
                        b.addStatement("%N.apply(_stmt, %L, %L)", mapper, sqlIndex, fieldValue)
                    }
                    sqlIndex++
                }
            }
        }
        if (batchParam != null) {
            b.endControlFlow()
        }
    }
}
